WordPress/gutenberg
# Add new TreeGrid and RovingTabIndex components

## Description
Extracted from PR #18014

This PR adds new components to the `block-editor` package for implementing a tree grid as described at [WAI ARIA authoring practices - Tree Grid](https://www.w3.org/TR/wai-aria-practices/examples/treegrid/treegrid-1.html).

I've created this as a separate PR as I think the implementation should get some scrutiny.

The new components:
- `TreeGrid`, `TreeGridRow`, `TreeGridCell` - implements keyboard navigation as described in the above link. up/down arrow keys navigate between grid rows, left/right arrows navigate between focusable elements in a row.
- `RovingTabIndex` a component that wraps a UI control that requires a roving tab index. This component  keeps track of the last focused element in the structure it wraps.
- `RovingTabIndexItem` a component that wraps an individual element/component that is focusable and requires a roving tab index. This component sets the correct `tabIndex` on the element/component.

## How has this been tested?
Testing can be carried out on PR #18014. Testing instructions:
1. Add a Navigation Block
2. Add some Navigation Links to the Navigation Block
3. Access the Block Navigator from the Navigation Block's toolbar
4. Tab into the navigation list
5. Tab again—observe the list only has a single tab stop
6. Shift tab back into the list
7. Use arrow keys to navigate rows and buttons within the rows
8. Shift tab out of the list
9. Tab back into the list—observe the last focused element in the list receives focus again.


## Types of changes


New feature (non-breaking change which adds functionality)


## Checklist:
